src/policy/mallocspace/global.rs

```diff
@@ -214,45 +249,180 @@ impl<VM: VMBinding> MallocSpace<VM> {
         if object.is_null() {
             return object;
         }
+
         let address = object.to_address();
         assert!(
             self.in_space(object),
             "Cannot mark an object {} that was not alloced by malloc.",
             address,
         );
-        if !is_marked::<VM>(object) {
-            set_mark_bit::<VM>(object);
+
+        if !is_marked::<VM>(object, None) {
+            let chunk_start = conversions::chunk_align_down(address);
+            set_mark_bit::<VM>(object, Some(Ordering::SeqCst));
+            set_chunk_mark(chunk_start);
             trace.process_node(object);
         }
+
         object
     }
 
-    pub fn release_all_chunks(&self) {
-        let mut released_chunks: HashSet<Address> = HashSet::new();
+    fn map_meta_space_for_chunk(&self, chunk_start: Address) {
+        // Map the metadata space for chunk
+        map_meta_space_for_chunk(&self.metadata, chunk_start);
+
+        // Update the bounds of the max and min chunk addresses seen -- this is used later in the sweep
+        // Lockless compare-and-swap loops perform better than a locking variant
+        let chunk_usize = chunk_start.as_usize();
+        let mut min = self.chunk_addr_min.load(Ordering::Relaxed);
+        let mut max = self.chunk_addr_max.load(Ordering::Relaxed);
+
+        while chunk_usize < min {
+            match self.chunk_addr_min.compare_exchange_weak(
+                min,
+                chunk_usize,
+                Ordering::AcqRel,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => break,
+                Err(x) => min = x,
+            }
+        }
+
+        while chunk_usize > max {
+            match self.chunk_addr_max.compare_exchange_weak(
+                max,
+                chunk_usize,
+                Ordering::AcqRel,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => break,
+                Err(x) => max = x,
+            }
+        }
+    }
+
+    pub fn sweep_chunk(&self, chunk_start: Address) {
+        // Call the relevant sweep function depending on the location of the mark bits
+        match VM::VMObjectModel::LOCAL_MARK_BIT_SPEC {
+            MetadataSpec::OnSide(local_mark_bit_side_spec) => {
+                unsafe { self.sweep_chunk_mark_on_side(chunk_start, local_mark_bit_side_spec) };
```

Comment:
I think marking both `sweep_chunk_mark_on_side()` and `sweep_chunk_mark_in_header()` functions as `unsafe` is pointless.
Rather, we should narrow-down the unsafety to more specific areas.

Non-atomic side metadata access functions are good examples of using unsafe for a whole function.
There, we have the atomic access functions which are the safer alternatives. So, there are two points in declaring non-atomic functions as unsafe:
1 - atomic access is a safer and expected alternative,
2 - non-atomic access functions only wrap around an unsafe non-atomic operation on an `Address`, which means we are not widening the unsafe scope.

In the case of `sweep_chunk_mark_on_side()` and `sweep_chunk_mark_in_header()` functions, none of the above reasons hold true.
Most importantly, there are several locations in these functions where unsafe operations are being used.
So, we are widening the unsafe scope.
Additionally, the caller of these functions can not do anything about their unsafety (while for non-atomic side metadata access functions, they have the atomic alternative).

Reply:
@qinsoon,
What do you think about these cases?

Reply:
Alright -- I'll reduce the unsafe scope.